core/Updates/4.3.0-b2.php

```diff
@@ -0,0 +1,55 @@
+<?php
+/**
+ * Matomo - free/libre analytics platform
+ *
+ * @link https://matomo.org
+ * @license http://www.gnu.org/licenses/gpl-3.0.html GPL v3 or later
+ *
+ */
+
+namespace Piwik\Updates;
+
+use Piwik\Updater;
+use Piwik\Updates as PiwikUpdates;
+use Piwik\Updater\Migration;
+use Piwik\Updater\Migration\Factory as MigrationFactory;
+use Piwik\Db;
+use Piwik\Common;
+use Piwik\Log;
+
+/**
+ * Update for version 4.3.0-b2.
+ */
+class Updates_4_3_0_b2 extends PiwikUpdates
+{
+    /**
+     * @var MigrationFactory
+     */
+    private $migration;
+
+    public function __construct(MigrationFactory $factory)
+    {
+        $this->migration = $factory;
+    }
+
+    public function getMigrations(Updater $updater)
+    {
+        $migrations = [];
+
+        $migrations[] = $this->migration->db->addColumn('segment', 'hash', 'CHAR(32) NULL', 'definition');
+
+        $segmentTable = Common::prefixTable('segment');
+        $segments = Db::fetchAll('SELECT idsegment, definition from ' . $segmentTable);
+        foreach ($segments as $segment) {
+            $hash = md5(urlencode($segment['definition']));
```

Comment:
This should be `md5($segment['definition'])`. Apologies, I know I said it needs to be `urlencode`d, I was referring to Segment construction. The `Segment::getSegmentHash()` function decodes the definition before `md5`-ing, so:

```
$segment = new Segment(urlencode($segment['definition']));
$hash = $segment->getHash();
```

is the same as

```
md5(urldecode(urlencode($segment['definition'])))
```

So if not going through `Segment`, calling `md5` directly works (but only for definitions in the segment table).

Reply:
@diosmosis and @tsteur there is an issue with this one.
If we have a segment currently in a db like this `actions%3E%3D2`, current Matomo version will use the hash of the `urlencoded` version of it, so when we create the hash here:

```php
$hash = md5($segment['definition']);
```

it won't be the same as we have in the `archive_numeric` tables.
One solution would be using `$hash = md5(urldecode($segment['definition']));` in the update file. 
We don't support double encoded segments in the db, do we? Only in URL/API?

Reply:
The segment condition value is encoded multiple times, not the segment condition operator. 

The hash is computed overall with the fully urlencoded segment, because that is what is sent in the `segment` parameter. The stored segments are created by sending a `definition=` parameter, which does not have special treatment and gets urldecoded once automatically. So there is a discrepancy with what is in the stored segment table and what is sent as a query parameter (`segment=urlencode(pageUrl==urlencode(urlencode(segmentValue)))`. The hash in the table has to match the hash computed w/ the fully encoded segment. The tests should cover this.

Reply:
The hash that gets put in the table needs to be the same hash that is created in this test: https://github.com/matomo-org/matomo/blob/4.x-dev/plugins/SegmentEditor/tests/System/ApiTest.php#L21 when run on 4.x-dev.

Reply:
thanks, @diosmosis, much more clear now.